### ISSUE TITLE
KISS-GP and Gradients

### ISSUE THREAD

**Issue by Thavisha** [cornellius-gp/gpytorch#1353]
I'm trying to use KISS-GP with gradient constraints (using RBFKernelGrad) for a 3-dimensional function, however there's no documentation to cover any cases of KISS-GP + derivatives, so I'm not quite sure if I'm on the right track here. My model currently looks like

`
class exactGP_kissgp(gpytorch.models.ExactGP):

     def __init__(self, train_x, train_y, likelihood):

             super(exactGP_kissgp, self).__init__(train_x, train_y, likelihood)

             grid_size = gpytorch.utils.grid.choose_grid_size(train_x,1.0)

             self.mean_module = gpytorch.means.ConstantMeanGrad()

             self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.GridInterpolationKernel(
                                                            gpytorch.kernels.RBFKernelGrad(ard_num_dims=3), grid_size=grid_size, num_dims=3,
                                                                                                ),
                                                            lengthscale_constraint=gpytorch.constraints.Interval(lower_bound=10, upper_bound=50) 
                                                        )                                                                        

    def forward(self, x):

        mean_x = self.mean_module(x)

        covar_x = self.covar_module(x) 

        return gpytorch.distributions.MultitaskMultivariateNormal(mean_x, covar_x)`

But when I attempt to predict I get the following error:

`
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/models/exact_gp.py", line 328, in __call__
    predictive_mean, predictive_covar = self.prediction_strategy.exact_prediction(full_mean, full_covar)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/models/exact_prediction_strategies.py", line 313, in exact_prediction
    test_test_covar = joint_covar[..., self.num_train :, self.num_train :]
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/lazy/lazy_evaluated_kernel_tensor.py", line 329, in __getitem__
    return self._getitem(row_index, col_index, *batch_indices)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/lazy/lazy_evaluated_kernel_tensor.py", line 56, in _getitem
    return self.evaluate_kernel()._getitem(row_index, col_index, *batch_indices)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/utils/memoize.py", line 51, in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/lazy/lazy_evaluated_kernel_tensor.py", line 274, in evaluate_kernel
    res = self.kernel(x1, x2, diag=False, last_dim_is_batch=self.last_dim_is_batch, **self.params)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/kernel.py", line 396, in __call__
    res = lazify(super(Kernel, self).__call__(x1_, x2_, last_dim_is_batch=last_dim_is_batch, **params))
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/module.py", line 28, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/scale_kernel.py", line 92, in forward
    orig_output = self.base_kernel.forward(x1, x2, diag=diag, last_dim_is_batch=last_dim_is_batch, **params)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/grid_interpolation_kernel.py", line 177, in forward
    base_lazy_tsr = lazify(self._inducing_forward(last_dim_is_batch=last_dim_is_batch, **params))
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/grid_interpolation_kernel.py", line 143, in _inducing_forward
    return super().forward(self.grid, self.grid, last_dim_is_batch=last_dim_is_batch, **params)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/grid_kernel.py", line 139, in forward
    covars = delazify(self.base_kernel(first_grid_point, full_grid, last_dim_is_batch=True, **params))
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/kernel.py", line 396, in __call__
    res = lazify(super(Kernel, self).__call__(x1_, x2_, last_dim_is_batch=last_dim_is_batch, **params))
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/module.py", line 28, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/opt/anaconda3/lib/python3.7/site-packages/gpytorch/kernels/rbf_kernel_grad.py", line 84, in forward
    K[..., :n1, :n2] = K_11
RuntimeError: expand(torch.FloatTensor{[3, 1, 10]}, size=[1, 10]): the number of sizes provided (2) must be greater or equal to the number of dimensions in the tensor (3)`

It's not clear to me exactly what the problem is. Playing with the dimensions of the kernels produced other error messages, but they seemed to make sense and imply that the dimensionality of the model is appropriate. Is this a fundamental limitation (as in, I can't use KISS-GP with RBFKernelGrad) or is there something I'm missing?

**Comment by FuhgJan** [comment]
Have you figured it out by now? I have been stuck on the same issue.

**Comment by Thavisha** [comment]
No unfortunately  not yet.

**Comment by gpleiss** [comment]
Sorry for the slow reply. Are you trying to implement what's described in [Scaling Gaussian Process Regression with Derivatives](https://papers.nips.cc/paper/2018/file/c2f32522a84d5e6357e6abac087f1b0b-Paper.pdf)?

`GridInterpolationKernel` is not compatible with the derivative kernels. At the moment, I don't think there is a way to combine grid interpolation with derivative kernels. (@dme65 ?)

**Comment by dme65** [comment]
> Sorry for the slow reply. Are you trying to implement what's described in [Scaling Gaussian Process Regression with Derivatives](https://papers.nips.cc/paper/2018/file/c2f32522a84d5e6357e6abac087f1b0b-Paper.pdf)?
> 
> `GridInterpolationKernel` is not compatible with the derivative kernels. At the moment, I don't think there is a way to combine grid interpolation with derivative kernels. (@dme65 ?)

Yeah, I don't think there is way to do that currently. We will have to generalize the grid kernel to support dealing with the block-wise Kronecker/Toeplitz structures as well as differentiating the SKI weights correctly if you use this in interpolation mode.

**Comment by Thavisha** [comment]
> Sorry for the slow reply. Are you trying to implement what's described in [Scaling Gaussian Process Regression with Derivatives](https://papers.nips.cc/paper/2018/file/c2f32522a84d5e6357e6abac087f1b0b-Paper.pdf)?

Yes, I think that's what I was aiming for.

> 
> `GridInterpolationKernel` is not compatible with the derivative kernels. At the moment, I don't think there is a way to combine grid interpolation with derivative kernels. (@dme65 ?)
 
> Yeah, I don't think there is way to do that currently. We will have to generalize the grid kernel to support dealing with the block-wise Kronecker/Toeplitz structures as well as differentiating the SKI weights correctly if you use this in interpolation mode.

Thanks for describing the issue for me. It's a shame it doesn't work yet. I guess I will shelve this problem for now and see if I can find an alternative approach. If it becomes possible in future I'll give it a try again. Thanks!
